src/chol_derivative.h

```diff
@@ -0,0 +1,183 @@
+    }
+};
+
+struct chol_jacobian {
+  int dim_cov_mat;
+  string cov_type;
+  chol mychol;
+  chol_jacobian(int dim, string cov): dim_cov_mat(dim), cov_type(cov), mychol(dim, cov) {};
+  template<class T>
+    vector<T> operator() (vector<T> &theta) {
+      return autodiff::jacobian(this->mychol, theta).vec();
+    }
+};
+
+vector<double> as_vector(NumericVector input);
+vector<int> as_vector(IntegerVector input);
+NumericVector as_nv(vector<double> input);
+NumericMatrix as_mv(matrix<double> input);
+matrix<double> as_matrix(NumericMatrix input);
+
+template <class Type>
+std::map<std::string, matrix<Type>> derivatives(int n_visits, std::string cov_type, vector<Type> theta) {
+  std::map<std::string, matrix<Type>> ret;
+  chol chol_obj(n_visits, cov_type);
+  chol_jacobian chol_jac_obj(n_visits, cov_type);
+  matrix<Type> l = chol_obj(theta).matrix();
+  l.resize(n_visits, n_visits);
+  ret["chol"] = l;
+  vector<Type> chol_d1_vec = autodiff::jacobian(chol_obj, theta).vec(); // chol_d1_vec is (dim * dim * l_theta)
+  vector<Type> chol_d2_vec = autodiff::jacobian(chol_jac_obj, theta).vec(); // chol_d2_vec is (dim * dim * l_theta * l_theta)
+  matrix<Type> ret_d1 = matrix<Type>(n_visits * theta.size(), n_visits);
+  matrix<Type> ret_d2 = matrix<Type>(n_visits * theta.size() * theta.size(), n_visits);
+  int n_visits_sq = n_visits * n_visits;
+  for (int i = 0; i < theta.size(); i++) {
+    matrix<Type> ld1 = chol_d1_vec.segment(i * n_visits_sq, n_visits_sq).matrix();
+    ld1.resize(n_visits, n_visits);
+    matrix<Type> ld1_lt = ld1 * l.transpose();
+    auto sigma_d1_i = ld1_lt + ld1_lt.transpose();
+    ret_d1.block(i * n_visits, 0, n_visits, n_visits) = sigma_d1_i;
+    for (int j = 0; j < theta.size(); j++) {
+      matrix<Type> ld2 = chol_d2_vec.segment( (j * theta.size() + i) * n_visits_sq, n_visits_sq).matrix();
+      matrix<Type> ld1_j = chol_d1_vec.segment(j * n_visits_sq, n_visits_sq).matrix();
+      ld2.resize(n_visits, n_visits);
+      ld1_j.resize(n_visits, n_visits);
+      auto ld2_lt = ld2 * l.transpose();
+      auto ld1_ld1j = ld1 * ld1_j.transpose();
+      auto sigma_d2_ij = ld2_lt + ld2_lt.transpose() + ld1_ld1j + ld1_ld1j.transpose();
+      ret_d2.block((i * theta.size() + j) * n_visits, 0, n_visits, n_visits) = sigma_d2_ij;
+    }
+  }
+  ret["derivative1"] = ret_d1;
+  ret["derivative2"] = ret_d2;
+  return ret;
+}
+
+template <class Type>
+struct chols {
+  std::map<std::vector<int>, matrix<Type>> inverse_cache;
+  std::map<std::vector<int>, matrix<Type>> sigmad1_cache;
+  std::map<std::vector<int>, matrix<Type>> sigmad2_cache;
+  std::map<std::vector<int>, matrix<Type>> sigma_cache;
+  std::map<std::vector<int>, matrix<Type>> sigma_inverse_d1_cache;
+  std::map<std::vector<int>, Eigen::SparseMatrix<Type>> sel_mat_cache;
+  std::string cov_type;
+  std::vector<int> full_visit;
+  int n_visits;
+  int n_theta; // theta.size()
+  vector<Type> theta;
+  matrix<Type> chol_full;
+  chols() {
+    
+  }
```

Comment:
correct?

Reply:
this is created to enable [] methods (it require default constructor), we can remove it and instead of using `[]` we use `insert` and `get` functionalities. This is added just to create a default initializer without initializing anything

Reply:
Ok better add a comment to explain for developers later